imported/w3c/web-platform-tests/navigation-api/navigation-methods/back-forward-multiple-frames.html is failing

cdumez opened the pull request:
#### 45eb3846b6f4bdd85e0e7181411942dd4c71960c
<pre>
imported/w3c/web-platform-tests/navigation-api/navigation-methods/back-forward-multiple-frames.html is failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=284865">https://bugs.webkit.org/show_bug.cgi?id=284865</a>

Reviewed by Rob Buis.

When calling `navigation.back()` / `navigation.forward()`, we would call `Page::goToItem()` which
would navigate a single frame. However, per the HTML specification, several frames may get navigated.
In particular, if a frame get navigated within the same document (e.g. fragment navigation), its
descendant frames may get navigated to whatever their URL was when their ancestor last was on this
HistoryItem.

In particular, the &quot;apply the history step&quot; [1] says:
&quot;&quot;&quot;
Let changingNavigables be the result of get all navigables whose current session history entry will change or reload given traversable and targetStep.
&quot;&quot;&quot;
Which refers to the algorithm at [2] to gather the changing navigables.

[1] <a href="https://html.spec.whatwg.org/#apply-the-history-step">https://html.spec.whatwg.org/#apply-the-history-step</a>
[2] <a href="https://html.spec.whatwg.org/#get-all-navigables-whose-current-session-history-entry-will-change-or-reload">https://html.spec.whatwg.org/#get-all-navigables-whose-current-session-history-entry-will-change-or-reload</a>

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigation-traverseTo-in-iframe-same-document-preventDefault-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigation-traverseTo-navigates-top-and-same-doc-child-and-cross-doc-child-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/back-forward-multiple-frames-expected.txt:
Rebaseline tests that are now passing or running further.

* Source/WebCore/bindings/js/JSDOMPromiseDeferred.h:
(WebCore::DeferredPromise::wasRejected const):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::goToItem):
(WebCore::HistoryController::goToItemForNavigationAPI):
(WebCore::HistoryController::recursiveGatherFramesToNavigate):
(WebCore::HistoryController::recursiveSetProvisionalItem):
(WebCore::HistoryController::itemsAreClones):
(WebCore::HistoryController::itemsAreClones const): Deleted.
* Source/WebCore/loader/HistoryController.h:
* Source/WebCore/loader/NavigationScheduler.cpp:
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::goToItemForNavigationAPI):
* Source/WebCore/page/Page.h:

Canonical link: <a href="https://commits.webkit.org/288123@main">https://commits.webkit.org/288123@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4cb34e96e90d6d45b4b8454b2f2094c5ff771345

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81927 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1454 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35884 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86484 "") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/32958 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84033 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1489 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9277 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/86484 "") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/32958 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84997 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/1109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74554 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/86484 "") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1011 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28731 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31378 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72340 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29340 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87915 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9169 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/72253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9354 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70373 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/71475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17814 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15579 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/14500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9120 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14656 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8960 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12485 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10768 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/bindings/js/JSDOMPromiseDeferred.h:
```diff
@@ -164,6 +164,12 @@ class DeferredPromise : public DOMGuarded<JSC::JSPromise> {
+    bool wasRejected() const
+    {
+        auto* lexicalGlobalObject = globalObject();
+        return deferred()->status(lexicalGlobalObject->vm()) == JSC::JSPromise::Status::Rejected;
```
Do we need to ASSERT deferred() and globalObject() like in some other places?

cdumez replied:
I can add them.

rwlbuis commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -344,12 +344,92 @@ void HistoryController::goToItem(HistoryItem& targetItem, FrameLoadType type, Sh
+
+    // Set the BF cursor before commit, which lets the user quickly click back/forward again.
+    // - plus, it only makes sense for the top level of the operation through the frame tree,
+    // as opposed to happening for some/one of the page commits that might happen soon
```
Nit: needs to end with a period, period.

rwlbuis reviewed (commented):
LGTM.

cdumez committed:
imported/w3c/web-platform-tests/navigation-api/navigation-methods/back-forward-multiple-frames.html is failing
https://bugs.webkit.org/show_bug.cgi?id=284865

Reviewed by Rob Buis.

When calling `navigation.back()` / `navigation.forward()`, we would call `Page::goToItem()` which
would navigate a single frame. However, per the HTML specification, several frames may get navigated.
In particular, if a frame get navigated within the same document (e.g. fragment navigation), its
descendant frames may get navigated to whatever their URL was when their ancestor last was on this
HistoryItem.

In particular, the "apply the history step" [1] says:
"""
Let changingNavigables be the result of get all navigables whose current session history entry will change or reload given traversable and targetStep.
"""
Which refers to the algorithm at [2] to gather the changing navigables.

[1] https://html.spec.whatwg.org/#apply-the-history-step
[2] https://html.spec.whatwg.org/#get-all-navigables-whose-current-session-history-entry-will-change-or-reload

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigation-traverseTo-in-iframe-same-document-preventDefault-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigation-traverseTo-navigates-top-and-same-doc-child-and-cross-doc-child-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/back-forward-multiple-frames-expected.txt:
Rebaseline tests that are now passing or running further.

* Source/WebCore/bindings/js/JSDOMPromiseDeferred.h:
(WebCore::DeferredPromise::wasRejected const):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::goToItem):
(WebCore::HistoryController::goToItemForNavigationAPI):
(WebCore::HistoryController::recursiveGatherFramesToNavigate):
(WebCore::HistoryController::recursiveSetProvisionalItem):
(WebCore::HistoryController::itemsAreClones):
(WebCore::HistoryController::itemsAreClones const): Deleted.
* Source/WebCore/loader/HistoryController.h:
* Source/WebCore/loader/NavigationScheduler.cpp:
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::goToItemForNavigationAPI):
* Source/WebCore/page/Page.h:

Canonical link: https://commits.webkit.org/288123@main
